design/sync-labels-bmh-to-node.md

```diff
@@ -0,0 +1,353 @@
+with the labels on the Node object. For example,
+
+```yaml
+    kind: BareMetalHost
+    name: node-0
+    metadata:
+      labels:
+        my-prefix.metal3.io/rack: xyz-123
+        my-prefix.metal3.io/zone: security-level-0
+        some-other-prefix.metal3.io/foo: bar
+    ...
+
+    ---
+    kind: Node
+    name: worker-node-0
+    metadata:
+      labels:
+        my-prefix.metal3.io/rack: xyz-123
+        my-prefix.metal3.io/zone: security-level-0
+    …
+
+```
+
+It's assumed that labels beginning with the specified prefix(es) are owned by
+CAPM3 and CAPM3 is solely responsible for keeping the set of labels in sync
+between BMH and Node objects. In other words, for the given set of label
+prefixes, the BMH is the source of truth and CAPM3 will enforce their
+presence on the Node.
+
+Additionally, there are certain well-known prefixes (e.g.
+kubelet.kubernetes.io/*, beta.kubernetes.io/*, kubernetes.io/*, k8s.io/*) that
+should be disallowed. Note the distinction between *.kubernetes.io/* (allowed)
+versus kubernetes.io/* (disallowed).
+
+### Synchronization Approach
+
+Synchronization must handle two scenarios: (i) a label is added/removed on a
+BMH and must be added/removed on the corresponding Node and (ii) a label is
+added/removed on the Node and must be removed/re-added on the Node to bring it
+in sync with the labels on BMH.
+
+The actual implementation of the synchronization logic may be encapsulated
+into an additional controller running within CAPM3; however, in the discussion
+below we simply reference CAPM3 as the entity doing the work.
+
+Scenario (i) can be handled as follows:
+
+1. CAPM3 will set up a watch on the BMH resource. It only requires read-only
+  access similar to what is done today:
+
+    baremetalhosts,verbs=get;list;watch
+
+2. On each BMH reconcile event, the controller will fetch the corresponding
+  Node. The _mapping_ can be accomplished by:
+
+    BMH.ConsumerRef --> Metal3Machine.OwnerRef --> Machine.NodeRef
+
+3. Synchronize the labels and update the Node in the workload cluster. CAPM3
+  already fetches the Node objects from the workload clusters today using a
+  remote client. We would leverage that same capability here.
+
+An alternative means to accomplish the mapping is to use the ProviderID:
+
+    BMH.ConsumerRef --> Metal3Machine.Spec.ProviderID --> Node.Spec.ProviderID
+
+Unless we can set up filters, the approach would require us to fetch all Nodes
+and then scan through them to find the Node matching the ProviderID. This does
+not seem scalable for large clusters.
+Field-selectors could potentially be employed and would require further
+investigation. However,there is no obvious advantage over the earlier approach
+for doing this mapping.
+
+Scenario (ii) can be handled with a full-sync of all BMH and Node objects at a
+preconfigured synchronization interval:
+
+1. At a configurable interval (e.g. 15 seconds), fetch complete list of BMHs.
```

Comment:
Instead of looping and fetching a list, it could watch hosts and then always return a request to re-enqueue with a delay. That would take advantage of the controller runtime logic for managing the work queue and keep the Reconcile() function a bit simpler.